### PR TITLE
[W10H02] First Commit

### PR DESCRIPTION
**Pull request by Anatoly03**
Tests for [W10H02](https://artemis.in.tum.de/courses/201/exercises/8841)

- `ArtemisTest.java` Tests following Artemis Examples
- `Lib.java` Reflections for getting private fields and static penguins.
- `UnitTest.java` Tests not from Artemis using extra special pingus.

If you want to add yourself as a penguin do it in `Lib.java`

### CODE REVIEW

**Review by DominikGleissner — APPROVED** [review]

File: w10h02/test/pgdp/teams/ArtemisTest.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+    @Test
+    @DisplayName("computeOptimalLineup: kleines Beispiel")
+    public void computeOptimalLineupSmall() throws NoSuchFieldException, IllegalAccessException {
+        var lineup = Lineup.computeOptimalLineup(Set.of(eric, nils, felix, thomas), 2, 1, 1);
+
+        var attackers = getField(lineup, "attackers");
+        var defenders = getField(lineup, "defenders");
+        var supporters = getField(lineup, "supporters");
+
```
Comment (Friendly-Banana) [inline w10h02/test/pgdp/teams/ArtemisTest.java:43]: That could be simplified, declare the expected result and then compare:

```java
Set attackersExpected = Set.of(nils, eric);
Set defendersExpected = Set.of(felix);
Set supportersExpected = Set.of(thomas);

assertEquals(attackersExpected, attackers);
assertEquals(defendersExpected, defenders);
assertEquals(supportersExpected, supporters);
```

File: w10h02/test/pgdp/teams/ArtemisTest.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        long timeAfter = System.currentTimeMillis();
+        long elapsedTime = timeAfter - timeBefore;
+
+        System.out.printf("Test `computeOptimalLineupLarge()` took %sms to execute.\n", elapsedTime);
+
+        var attackers = getField(lineup, "attackers");
+        var defenders = getField(lineup, "defenders");
+        var supporters = getField(lineup, "supporters");
+
+        assertEquals(2, attackers.size());
```
Comment (Friendly-Banana) [inline w10h02/test/pgdp/teams/ArtemisTest.java:73]: see above

File: w10h02/test/pgdp/teams/UnitTest.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    @Test
+    @DisplayName("Die Summe der Parameter numberXXX der Methode computeOptimalLineup ist <= der Anzahl der übergebenen Penguins (players.size())")
+    public void optimalLineupAllAttackers() throws NoSuchFieldException, IllegalAccessException {
+        var lineup = Lineup.computeOptimalLineup(Set.of(glen, fatjon, jani, koco), 4, 0, 0);
+
+        var attackers = getField(lineup, "attackers");
+        var defenders = getField(lineup, "defenders");
+        var supporters = getField(lineup, "supporters");
+
+        assertEquals(4, attackers.size());
```
Comment (Friendly-Banana) [inline w10h02/test/pgdp/teams/UnitTest.java:46]: see above

File: w10h02/test/pgdp/teams/Lib.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    //
+    // Disclaimer: Ähnlichkeiten zu realen Pinguinen sind rein zufällig.
+    //
+
+    // Artemis Example 1
+    public static Penguin jonas = new Penguin("Jonas", 10, 0, 0);
+    public static Penguin anatoly = new Penguin("Anatoly", 10, 10, 0);
+    public static Penguin julian = new Penguin("Juilan", 10, 10, 0);
```
Comment (Friendly-Banana) [inline w10h02/test/pgdp/teams/Lib.java:23]: Did you mean `Julian`?

**Review by Friendly-Banana — CHANGES_REQUESTED** [review]
Small improvements, otherwise LGTM.

File: w10h02/test/pgdp/teams/Lib.java (hunk 4 above)
Comment (Anatoly03) [reply w10h02/test/pgdp/teams/Lib.java:23]: No, I’m following Artemis Typos. Blame not me.

File: w10h02/test/pgdp/teams/ArtemisTest.java (hunk 1 above)
Comment (Anatoly03) [reply w10h02/test/pgdp/teams/ArtemisTest.java:43]: Do you mean `assertArrayEquals`?

Comment (Anatoly03) [reply w10h02/test/pgdp/teams/ArtemisTest.java:43]: Isn’t assertEquals per Reference?

**Comment by Anatoly03** [comment]
Would suggest to push this as per Dominik’s approval and edit in another PR

File: w10h02/test/pgdp/teams/ArtemisTest.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        assertEquals(1, defenders.size());
+        assertEquals(1, supporters.size());
+
+        assertTrue(attackers.contains(nils));
+        assertTrue(attackers.contains(eric));
+        assertTrue(defenders.contains(felix));
+        assertTrue(supporters.contains(thomas));
```
Comment (Friendly-Banana) [inline w10h02/test/pgdp/teams/ArtemisTest.java:51]: Set attackersExpected = Set.of(nils, eric);
Set defendersExpected = Set.of(felix);
Set supportersExpected = Set.of(thomas);

assertEquals(attackersExpected, attackers);
assertEquals(defendersExpected, defenders);
assertEquals(supportersExpected, supporters);

Comment (Anatoly03) [reply w10h02/test/pgdp/teams/ArtemisTest.java:51]: Use ```suggest

I think

Comment (Friendly-Banana) [inline w10h02/test/pgdp/teams/ArtemisTest.java:51]: Set attackersExpected = Set.of(nils, eric);
Set defendersExpected = Set.of(felix);
Set supportersExpected = Set.of(thomas);

assertEquals(attackersExpected, attackers);
assertEquals(defendersExpected, defenders);
assertEquals(supportersExpected, supporters);

File: w10h02/test/pgdp/teams/ArtemisTest.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+        assertTrue(attackers.contains(max));
+        assertTrue(attackers.contains(georg));
+        assertTrue(defenders.contains(jan));
+        assertTrue(defenders.contains(anton));
+        assertTrue(defenders.contains(konrad));
+        assertTrue(supporters.contains(lukas));
+        assertTrue(supporters.contains(laura));
+        assertTrue(supporters.contains(johannes));
+        assertTrue(supporters.contains(oliver));
+        assertTrue(supporters.contains(robin));
```
Comment (Friendly-Banana) [inline w10h02/test/pgdp/teams/ArtemisTest.java:86]: Set attackersExpected = Set.of(max,georg);
Set defendersExpected = Set.of(jan,anton,konrad);
Set supportersExpected = Set.of(lukas,laura,johannes,oliver,robin);

assertEquals(attackersExpected, attackers);
assertEquals(defendersExpected, defenders);
assertEquals(supportersExpected, supporters);

File: w10h02/test/pgdp/teams/UnitTest.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        assertEquals(0, defenders.size());
+        assertEquals(0, supporters.size());
+
+        assertTrue(attackers.contains(glen));
+        assertTrue(attackers.contains(fatjon));
+        assertTrue(attackers.contains(jani));
+        assertTrue(attackers.contains(koco));
```
Comment (Friendly-Banana) [inline w10h02/test/pgdp/teams/UnitTest.java:53]: Set attackersExpected = Set.of(glen,fatjon,jani,koco);
        Set defendersExpected = Set.of();
        Set supportersExpected = Set.of();
    
        assertEquals(attackersExpected, attackers);
        assertEquals(defendersExpected, defenders);
        assertEquals(supportersExpected, supporters);

**Review by Friendly-Banana — CHANGES_REQUESTED** [review]

**Comment by Anatoly03** [comment]
@Friendly-Banana review please

**Review by Friendly-Banana — APPROVED** [review]

**Review by MaximilianAnzinger — APPROVED** [review]

**Pull request merged**